### PR TITLE
Extend GCP client with methods for extracting billing time series data from db

### LINKED ISSUE #27: Extend GCP client with methods for extracting billing time series data from db
- xproject/internal/cloud/gcpcln.
- extract raw or aggregate imformation by period (from, to) for using in main billing client.
- all access to db must be implemented from postgres client package.
- unit tests.

### PR DESCRIPTION
**Pull request by irodionr**
Resolves #27 
Adds new struct:
- `GcpAccount` - GCP user account information

Extends GCP client with following methods:
- `ListAccounts()` - returns all accounts
- `GetLastAccount()` - returns the latest account by ID
- `AddAccount(GcpAccount)` - adds account
- `ListFiles()` - returns all files' metadata
- `GetLastFile()` - returns the latest file by time
- `AddFile(gcptypes.FileMetadata)` - adds file
- `ListAllBills()` - returns all bills
- `ListBillsByTime(time.Time, time.Time)` - returns bills from specified time period
- `ListBillsByService(string)` - returns bills that are related to specified GCP service
- `ListBillsByProject(string)` - returns bills that are related to specified GCP project
- `GetLastBill()` - returns the latest added bill
- `AddBill(gcptypes.ServiceBill)` - adds bill
- `AddReport(gcptypes.Report)` - adds report
- `AddReportsToAccount(gcptypes.Reports, int)` - adds reports to specified account

### CODE REVIEW

File: internal/db/pgcln/client.go [hunk 1]
```diff
@@ -63,12 +105,461 @@ func New(conf Config) (*Client, error) {
...
+	return nil
+}
+
+// SelectBillsByTime returns bills from db that belong to specified time period
+func (c *Client) SelectBillsByTime(start, end time.Time) (ServiceBills, error) {
+	if start.After(end) || end.Before(start) {
+		return nil, fmt.Errorf("%v: invalid arguments err", pgcLogPref)
+	}
+
+	rows, err := c.idb.Query("SELECT * FROM xproject.service_bills ORDER BY id ASC")
```
Comment (sergusha) [inline internal/db/pgcln/client.go]: Why don't you SELECT only the needed data?
```sql
SELECT * FROM xporject.service_bills 
              WHERE start_time >= timestamp $1 
                   AND end_time < timestamp $2
         ORDER BY id ASC
```

Comment (irodionr) [reply internal/db/pgcln/client.go]: Fixed

File: internal/db/pgcln/client.go [hunk 2]
```diff
@@ -63,12 +105,457 @@ func New(conf Config) (*Client, error) {
...
+
+		table = append(table, &row)
+	}
+
+	return table, nil
+}
+
+// InsertAccount inserts account into db
+func (c *Client) InsertAccount(account Account) error {
+	if _, err := c.idb.Query("INSERT INTO xproject.accounts VALUES(DEFAULT, $1)",
```
Comment (qezz) [inline internal/db/pgcln/client.go]: The string queries are not so convenient to use, so we should firstly implement the Data Access Layer (#9) interface to use unified methods to receive the data from DBs.

Thus, we can avoid the repeated logic within methods.

File: internal/db/pgcln/client.go [hunk 3]
```diff
@@ -63,12 +105,457 @@ func New(conf Config) (*Client, error) {
...
+	result := make([]string, len(cols))
+	rawResult := make([][]byte, len(cols))
+	dest := make([]interface{}, len(cols))
+
+	for i := range rawResult {
+		dest[i] = &rawResult[i]
+	}
+
+	for rows.Next() {
+		if err := rows.Scan(dest...); err != nil {
```
Comment (qezz) [inline internal/db/pgcln/client.go]: I guess... there is a way to parse the returned values "on-the-fly"

Here (https://golang.org/pkg/database/sql/#Rows) is noted that `Scan()` can receive typed arguments, and fill them with correct values. This way, we don't need to parse the returned strings anymore.

File: internal/db/pgcln/client.go [hunk 4]
```diff
@@ -63,12 +105,457 @@ func New(conf Config) (*Client, error) {
...
+		log.Printf("%v: db query err, %v", pgcLogPref, err)
+		return err
+	}
+
+	return nil
+}
+
+// SelectCsvFiles returns all CSV files from db
+func (c *Client) SelectCsvFiles() (GcpCsvFiles, error) {
+	rows, err := c.idb.Query("SELECT * FROM xproject.gcp_csv_files ORDER BY id ASC")
```
Comment (qezz) [inline internal/db/pgcln/client.go]: I'm not sure, but you probably should `defer rows.Close()` after any `Query()` request.

File: internal/db/pgcln/client.go [hunk 5]
```diff
@@ -63,12 +101,199 @@ func New(conf Config) (*Client, error) {
...
+	}
+
+	return table, nil
+}
+
+// SelectLastBill returns the latest added bill from db
+func (c *Client) SelectLastBill() (ServiceBill, error) {
+	var row ServiceBill
+
+	rows, err := c.idb.Query("SELECT * FROM xproject.service_bills WHERE id = (SELECT MAX(id) FROM xproject.service_bills)")
```
Comment (rch9) [inline internal/db/pgcln/client.go]: I think `ID` can not guarantee right time order

Comment (rch9) [inline internal/db/pgcln/client.go]: I think ID can not guarantee right time order

File: internal/db/pgcln/client.go (hunk 2 above)
Comment (irodionr) [reply internal/db/pgcln/client.go]: Is this line of code supposed to look like
`if _, err := c.InsertInto("xproject.accounts", account.GcpAccountInfo)` ?
Where `InsertInto(string, string)` should be defined then?

**Comment by irodionr** [comment]
Travis needs to setup PostgreSQL database and create tables (https://github.com/pavlov-tony/xproject/blob/master/migrations/000001_initialize_xproject.up.sql) before unit testing.

File: internal/db/pgcln/client.go [hunk 6]
```diff
@@ -53,22 +60,279 @@ func New(conf Config) (*Client, error) {
...
+		log.Printf("%v: db query err, %v", pgcLogPref, err)
+		return nil, err
+	}
+	defer rows.Close()
+
+	return combineFiles(rows)
+}
+
+// GetLastFile returns last file's metadata from db by time
+func (c *Client) GetLastFile() (*gcptypes.FileMetadata, error) {
```
Comment (rch9) [inline internal/db/pgcln/client.go]: This method need not return error if Files table is empty

File: internal/db/pgcln/client.go [hunk 7]
```diff
@@ -53,22 +60,279 @@ func New(conf Config) (*Client, error) {
...
+		bill.FileMetadataID); err != nil {
+		log.Printf("%v: db query err, %v", pgcLogPref, err)
+		return err
+	}
+
+	return nil
+}
+
+// AddReport adds file's metadata and bills into db
+func (c *Client) AddReport(report gcptypes.Report) error {
```
Comment (rch9) [inline internal/db/pgcln/client.go]: I need also in method AddReports(gcptypes.Reports)

File: internal/db/pgcln/client_test.go [hunk 8]
```diff
@@ -0,0 +1,422 @@
...
+	"testing"
+	"time"
+
+	_ "github.com/lib/pq"
+	"github.com/pavlov-tony/xproject/pkg/cloud/gcptypes"
+)
+
+// TestGcpAccount tests adding account into db, listing and removing it
+func TestGcpAccount(t *testing.T) {
+	conf := Config{
```
Comment (pavlov-tony) [inline internal/db/pgcln/client_test.go]: consider that library for the sql mock purposes: https://github.com/DATA-DOG/go-sqlmock
Just be aware that in the real unit tests you should mock DB interactions.

File: internal/db/pgcln/client.go [hunk 9]
```diff
@@ -53,22 +60,333 @@ func New(conf Config) (*Client, error) {
...
+		log.Printf("%v: db query err, %v", pgcLogPref, err)
+		return nil, err
+	}
+	defer rows.Close()
+
+	return combineBills(rows)
+}
+
+// GetLastBill returns the latest added bill from db by time
+func (c *Client) GetLastBill() (*gcptypes.ServiceBill, error) {
```
[the remark this answers is not in the document]
Comment (irodionr) [reply internal/db/pgcln/client.go]: `gcptypes.ServiceBill` is a struct that might grow in the future, so it is accessed by pointer.
`gcptypes.ServicesBill` is a slice of pointers to `gcptypes.ServiceBill`, using a pointer to a slice is redundant and doesn't save more than a few bytes.

**Pull request merged**